face-detection-explainer.md

```diff
@@ -0,0 +1,481 @@
+
+## User research
+
+* TEAMS : Supportive, but [not supportive of adding facial expressions](https://www.w3.org/2021/11/24-webrtc-minutes.html#t04) and doubtful on the accuracy of emotion analysis.
+
+*Agreed, Facial expressions are not part of [Non-goals](https://github.com/riju/faceDetection/blob/main/explainer.md#non-goals) for this API*
+
+
+* MEET : Supportive, but many applications might not be using rectangle-bounding box, mask/contour more useful.
+
+*Currently common platforms such as ChromeOS, Android, and Windows support [system APIs](https://github.com/riju/faceDetection/blob/main/explainer.md#platform-support) which return only face bounding box and landmarks, not accurate contour, and therefore initial implementations are expected to support only bounding-box (ie. a contour with maximum of four points). We would keep the API extensible, so that proper contour support can be added in future. A few [use-cases for bounding-box face detection](https://github.com/riju/faceDetection/blob/main/explainer.md#key-scenarios) are listed.*
+
+
+* Zoom : 
+
+## Face Detection API
+
+### Common API
+
+```js
+dictionary DetectedFace {
+  required long                     id;
+  required float                    probability;
+  FrozenArray<Point2D>              contour;
+  FrozenArray<DetectedFaceLandmark> landmarks;
+};
+
+dictionary DetectedFaceLandmark {
+  required FrozenArray<Point2D> contour;
+  FaceLandmark                  type;
+};
+
+enum FaceLandmark {
+  "eye",
+  "eyeLeft",
+  "eyeRight",
+  "mouth",
+  "nose"
+};
+
+partial dictionary MediaTrackSupportedConstraints {
+  boolean faceDetectionMode = true;
+};
+
+partial dictionary MediaTrackCapabilities {
+  sequence<DOMString> faceDetectionMode;
+};
+
+partial dictionary MediaTrackConstraintSet {
+  ConstrainDOMString faceDetectionMode;
+};
+
+partial dictionary MediaTrackSettings {
+  DOMString faceDetectionMode;
+};
+
+enum FaceDetectionMode {
+  "none",         // Face detection is not needed
+  "presence",     // Only the presence of face or faces is returned, not location
+  "bounding-box", // Return bounding box for face
+  "contour",      // Approximate contour of the detected faces is returned
+  "landmarks",    // Approximate contour of the detected faces is returned with facial landmarks
+};
+```
+
+### Extended VideoFrame API (alternative 1)
+
+In this section we propose an API which extends WebCodecs `VideoFrame` and
+`VideoFrameMetadata` in order to provide face detection results.
+
+```js
+partial interface VideoFrame {
+  readonly attribute FrozenArray<DetectedFace>? detectedFaces;
+};
+
+partial dictionary VideoFrameMetadata {
+  FrozenArray<DetectedFace> detectedFaces;
+};
+```
+
+Example are shown in sections [Example 1-1](#example-1-1) and
+[Example 2-1](#example-2-1).
+
+### VideoFrame Side-Channel API (alternative 2)

```

Comment:
Maybe we should link to https://github.com/w3c/mediacapture-extensions/issues/70 and https://github.com/w3c/mediacapture-extensions/issues/70.

Reply:
I think the relevant link nowadays is the link to [VideoFrameMetadata](https://www.w3.org/TR/webcodecs/#dictdef-videoframemetadata) which was added into the explainer.